### PR TITLE
ToIterable execution in Beam runner

### PR DESCRIPTION
**Pull request by nownikhil**
We use the coder from the PCollection and store data as bytes in temp location using FileIO.
We later read this file using the same coder and get the data on the submitter node.
TESTS: Added unit tests + Tested with Dataflow Runner

### CODE REVIEW

File: scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamMode.scala
Comment (nownikhil) [inline scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamMode.scala:109]: Maybe a cleaner way to do this?

File: scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamMode.scala
Comment (johnynek) [inline scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamMode.scala]: seems like `1` should be a constructor param.

File: scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamMode.scala
Comment (johnynek) [inline scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamMode.scala:106]: is there no `close` in `FileIO.Sink`? Seems like we should have a way to close outputStream.

File: scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamMode.scala
Comment (johnynek) [inline scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamMode.scala]: this worries me. What if you see an error that is not EOF?

Also, boxing with Try is an extra cost you don't need.
```scala
try {
  nextRecord = coder.decode(stream)
  hasNextRecord = true
}
catch {
  case _: EOFException =>
    hasNextRecord = false
  // don't catch other errors.
}
```

File: scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala
Comment (johnynek) [inline scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala]: I think this needs to be thread-safe, so it may be best to use `ConcurrentHashMap` from java or use a mutex somehow.

File: scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala
Comment (johnynek) [inline scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala]: why do we enforce this? we could easily concatenate the Iterators right?

File: scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala
Comment (johnynek) [inline scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala]: seems like we should move this up to me:
```scala
Future {
  // Single dir by default just matches the dir, we need to match files inside
  val matchedResources = FileSystems.`match`(s"$path*").metadata()
  
  new Iterable[T] {
    def iterator = matchedResources.asScala.iterator.flatMap { m =>
      new InputStreamIterator(
        Channels.newInputStream(FileSystems.open(m.resourceId())),
        c)
    }
}
```

File: scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala
Comment (johnynek) [inline scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala]: this should probably be a `require`. An assertion is something that *must* be true by construction, not a requirement.

File: scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala
Comment (johnynek) [inline scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala:88]: just to be clear: this isn't blocking right? this adds it to a pipeline which is run later below right?

File: scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamMode.scala
Comment (johnynek) [inline scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamMode.scala]: you never use the `Pipeline` right? That is just because you are reusing the BeamSink interface? Does it need the pipeline?

File: scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala
Comment (johnynek) [inline scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala:84]: just to be clear: the above isn't blocking when it calls `run` or `write` is it? It kind of looks like it. I think we need a comment.

File: scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala
Comment (johnynek) [inline scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala]: this looks like it is blocking here. We don't want that. We want to run this concurrently with all the other operations I think. Can you add a comment?

I think the loop should build a set of actions that need to run before the future that returns the run ID an execution counters.

**Review by johnynek — COMMENTED** [review]
Also, we have no idea if the tests pass since the CI is disabled. I think we really need to get that working. It is very dangerous to merge PRs without a working CI.

File: scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamMode.scala
Comment (nownikhil) [reply scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamMode.scala:106]: It is handled by the underlying FileIO which uses FileIO.Sink

File: scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala
Comment (nownikhil) [reply scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala]: Done

File: scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala
Comment (nownikhil) [reply scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala:88]: It only builds up on the pipeline object which runs in the end.

File: scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamMode.scala
Comment (nownikhil) [reply scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamMode.scala]: Refactored the interface. Pipeline can be accessed from PCollection via `pcol.getPipeline`

File: scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala
Comment (nownikhil) [reply scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala]: Added comment

File: scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamMode.scala
Comment (johnynek) [inline scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamMode.scala]: nit: can we use our safer `widen` function here (general cast is more dangerous).

File: scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala
Comment (johnynek) [inline scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala]: nit: I think `matchedResources.iterator.flatMap { ` clearer than manually reimplementing via fold.

Now that I think about this though, we are never closing the input stream.

I think we may need something like this:

```scala
def iterator = {
  // an empty Iterator that closes an InputStream when it is iterated
  def closeIt(is: InputStream): Iterator[T] =
    new Iterator[T] {
      def hasNext = {
        is.close()
        false
      }
      def next = Iterator.empty.next
    }

  matchedResources
    .iterator
    .flatMap { resource =>
      val is = Channels.newInputStream(FileSystems.open(resource.resourceId()))
      (new InputStreamIterator(is, c)) ++ closeIt(is)
    }
}
```
At least that would close the resource when you iterate through all the items.

**Review by johnynek — COMMENTED** [review]
one concern about closing the InputStreams, but other than that I think it looks good.

That said, I will reiterate my concern about not having CI and having no idea if the tests pass or if this even compiles.

File: scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala
Comment (johnynek) [inline scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala:98]: force is almost the same as toIterable, you just materialize it and then make a PCollection to read it in the next pipeline, but you are smart to leave that to a follow up PR.

**Comment by nownikhil** [comment]
Someone from Scalding team wants to work on setting up CI for scalding repo. I will hold on to this patch till the CI is set up and apply it internally in source to test on real dataflow jobs. Once the jobs on dataflow succeed and CI is setup, I will merge this patch.

Tests are succeeding on my local. But this is not urgent to merge right now so better to wait here.

File: scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamMode.scala
Comment (nownikhil) [reply scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamMode.scala]: Done

File: scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala
Comment (nownikhil) [reply scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala:98]: Yup this will be done in a follow up patch

**Review by johnynek — APPROVED** [review]
sounds good!

File: scalding-beam/src/test/scala/com/twitter/scalding/beam_backend/BeamBackendTests.scala
Comment (johnynek) [inline scalding-beam/src/test/scala/com/twitter/scalding/beam_backend/BeamBackendTests.scala]: nit: `.mapValues { case (left, right) => left + right }` would be a bit more idiomatic (always prefer operating only on values if you can since you open up more opportunities for dag optimization.

**Comment by johnynek** [comment]
please merge master after #1965 is merged because scalding-beam was currently not being tested.

**Comment by johnynek** [comment]
(although master does pass for me, I assume this PR passes also).

**Pull request merged**
